Custom Colors

timschwab opened the pull request:
This revision enables the user to create a custom color palette. The effect can be seen [here](http://pyzaist.com/tri/dist/).

There is one known bug. If the user adds a palette with 0 or 1 colors, it gets added as usual, but the behavior is abnormal. If there are no colors, then the resulting trianglify canvas is all black. If there is only one color, an error is thrown and the trianglify canvas is not updated.

sschaub commented:
This looks really nice to me, Tim. 

On my laptop screen, the "Add Custom Palette" button you added leaves almost no room to see the color palettes. Part of the problem is the pre-release verbage at the top. But I'm wondering if you could move the "Add Custom Palette" button up so that it is level with the "Color" heading, and right-aligned with the variance and cell size values. That would give more space to see the color palettes, regardless of screen size.

In the color picker dialog, the "Add Palette" and "Cancel" buttons are jammed next to each other and need a bit of white space between.

Adding a custom palette with no colors should not be allowed. I suggest disabling the "Add Palette" button until at least two colors have been added. This would prevent the bug you mentioned above.

This is my $.02. I'm interested in what qrohlf thinks.

qrohlf commented on js/ColorItem.jsx:
```diff
@@ -0,0 +1,18 @@
+      </div>
+    );
+  }
+};
```
Nice extraction, this is much cleaner as a separate component.

qrohlf commented on js/ColorList.jsx:
```diff
@@ -1,29 +1,22 @@
 let React = require('react/addons');
+let ColorItem = require('./ColorItem.jsx');
+let CustomColor = require('./CustomColors.jsx');
```
For clarity's sake, can you rename this reference to `CustomColors` so that it matches the filename & component name?

qrohlf commented on js/CustomColors.jsx:
```diff
@@ -0,0 +1,84 @@
+
+          <br />
+
+          <input type="button" value="Add Palette" onClick={this.addPalette} />
```
disabling the input when the user hasn't added enough color (`disabled={this.state.palette.length < 3}`) will fix your known issues

qrohlf commented on js/Display.jsx:
```diff
@@ -1,6 +1,6 @@
 let React = require('react/addons');
 let TrianglifyCanvas = require('./TrianglifyCanvas.jsx');
-let Trianglify = require('Trianglify');
+let Trianglify = require('trianglify');
```
HFS+: because what could possibly go wrong with a case-insensitive, case-preserving filesystem? 
😞

qrohlf commented:
+1 on @sschaub's suggestion about moving the 'Add Custom Palette' button to the right of the 'Color' heading and the UI tweaks on the modal buttons. I also would suggest overriding the drop shadow from the `<ColorPicker.default/>` since it looks kind of weird to have a div with a drop shadow on top of a modal.

That said, the functionality here is spot-on, implemented nicely, and works well. A possible future enhancement would be to save the user-generated palettes in localstorage, but that's a PR for another day.

timschwab commented:
I added some comments, put some space between the buttons on the modal, and moved the button up. I also fixed the issues that were mentioned. I looked into removing the drop-shadow [here](https://github.com/casesandberg/react-color/issues/90), but it seemed like it would not be an easy fix. I think the only way to do it would be to make a custom color picker through the react-color api.

qrohlf commented:
Merged! I'm going to add a tooltip to the 'add palette' button when it's disabled and then I'll go ahead and deploy this change.

timschwab commented:
Great! Thank you for the pointers and being willing to work with me on the project.
